espnet2/asr/decoder/rnnt_decoder.py

```diff
@@ -0,0 +1,331 @@
+"""RNN-Transducer implementation for training and decoding."""
+
+import torch
+
+from typeguard import check_argument_types
+from typing import Dict
+from typing import List
+from typing import Tuple
+from typing import Union
+
+from espnet.nets.beam_search_transducer_espnet2 import Hypothesis
+from espnet.nets.pytorch_backend.nets_utils import get_activation
+from espnet.nets.pytorch_backend.nets_utils import to_device
+
+from espnet2.asr.decoder.abs_decoder import AbsDecoder
+
+
+class RNNTDecoder(AbsDecoder):
+    """RNN-T Decoder module.
```

Comment:
Can you consider RNNTDecoder into RNN module and joint module?
Otherwise, we need many many duplications only for transducer, RNNTDecoder, TransformerTDecoder, etc. This is one of the worst point of espnet1.

Reply:
Hum, I'm not sure I understand. You mean defining a general TransducerDecoder based on default `RNNDecoder` + `JointNetwork`? Sounds good to me but I'm not sure how to handle properly the different methods for RNN-T/RNN-T-Att

Reply:
Yes, exactly.

Reply:
OK, I'll do that! About the different methods between RNN-T/RNN-T-Att (`forward`, `score`, etc) what would be the correct approach in your opinion?

Reply:
Thank you.

Sorry,  but I didn't deeply think how to do. Probably, we can and need to create multiple JointNetworks for RNN-T/RNN-T-Att respectively.
Maybe, you can produce better idea.

Reply:
I made a new version, it could be improved but I guess it's a start.

Reply:
Sorry, my explanation might be bad. Your modification is against my intention. 

I don't think much yesterday. I'll say about ideal at the next. I'm not sure whether it's not possible.

Again. We must not create duplications as much as possible. This is more important than you may expected for the future development and maintainability. 
I requested you doing it with not a little effort to refactor the current transducer structure drastically.

1. We don't need a general `TransducerDecoder`.
2. We must **share the same decoders** between att/ctc hybrid and transducer. 

About training:

```python
# ESPNetASRModel.forward()
decoder_out = decoder()
joint_network(enoder_out, decoder_out)
```

About scoring:

Currently, the scoring methods are quite different between att/ctc hybrid and transducer. I'm not sure whether we can unify them. If it's impossible, then RNNDecoder should have both methods.

```python
# espnet2.asr.decoders.rnn_decoder
class RNNDecoder:
    def score(self, ...):
    def score_transducer(self, ...):

class TransducerScorerAdapter:
     def __init__(self, decoder):
           self.decoder
     def score(self, ...)
           self.decodoer.score_transducer(self, ...):

```

What do you think?

Reply:
> Again. We must not create duplications as much as possible. This is more important than you may expected for the future development and maintainability.

Oh I totally understand that for development and maintanaibility.

So, If we are clear here: you want me to extend `RNNDecoder` instead of inheriting `RNNDecoder` in `TransducerDecoder` and use `JointNetwork` only in `_calc_transducer_loss` for training? I'm also fine with that but I need to modify `RNNDecoder` to handle some cases (no attention mainly).
Outside `JointNetwork` I'll move, I'm not sure it would be better for maintainability? With some small modifications in current implementation (calling parent methods in `init_state` and `forward` when attention is used), most code would just be moved as there are not many duplications. It would also be harder in terms of readability I think?
 
> About scoring:
> 
> Currently, the scoring methods are quite different between att/ctc hybrid and transducer. I'm not sure whether we can unify them. If it's impossible, then RNNDecoder should have both methods.

I'm not sure either, I couldn't do it in my first try.
Btw, I'm planning to replace both transducer training and decoding part by a torch lib for multiple purposes (optimizations, new features, speed, etc) so unification will take another form I guess...